src/transformStateWithClones.js

```diff
@@ -7,7 +7,31 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateArray = [];
+  let newState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear':
+        newState = {};
+        break;
+
+      case 'addProperties':
+        Object.assign(newState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete newState[key];
+        }
+
+      default:
+        continue;
+    }
+    stateArray.push({ ...newState });
```

Comment:
Cloning the newState object inside the loop for each action can be inefficient if the state object is large. Consider if there's a way to handle the state changes that avoids unnecessary cloning or if cloning is indeed the best approach for each action.